[VDG] Wallet Settings + Coinjoin Profiles unification

ichthus1604 opened the pull request:
1 - Replaces ad-hoc controls in the Settings dialog by the Coinjoin Profile selector:

![image](https://user-images.githubusercontent.com/98904713/167344866-1bbe43a3-b96e-4dfb-82f4-090bdbf79d98.png)

2 - Adds a Slider to the Pleb Stop Threshold input and replaces the `TextBox` by a `CurrencyEntryBox`:

![image](https://user-images.githubusercontent.com/98904713/167345089-2915b2a9-4d40-41c6-9c6c-a688cf3d26a8.png)

This allows arbitrary user input as well as predefined slider values.

3 - The UI is now unified: these are the same dialogs and options you get when creating a new wallet.

MaxHillebrand commented:
Automatic coinjoin on/off should be on the first page, I think.

Also the text in customize is still horrible, we gotta rewrite to make it more understandable.

MaxHillebrand commented on WalletWasabi.Fluent/Views/Dialogs/ManualCoinJoinProfileDialogView.axaml:
```diff
@@ -13,26 +13,51 @@
+      </StackPanel>
+
+      <StackPanel Spacing="10" ToolTip.Tip="Minimum anonymity score for a coin to be considered private.">
         <TextBlock Text="Privacy Threshold" />
```
I think we need a different name here

MaxHillebrand commented on WalletWasabi.Fluent/Views/Dialogs/ManualCoinJoinProfileDialogView.axaml:
```diff
@@ -13,26 +13,51 @@
 
-      <StackPanel Spacing="10">
+      <StackPanel Spacing="10"  ToolTip.Tip="A coin will not be automatically registered for coinjoin after this anonymity score is reached.">
         <TextBlock Text="Privacy Sanity Check" />
```
Here again we need something better.


```suggestion
         <TextBlock Text="Max anonscore" />
```

ichthus1604 replied:
@MaxHillebrand can you suggest proper wording for this?

MarnixCroes commented:
When using `Cancel` it should not keep a previous selected item selected. Now it has `Maximize Privacy` selected even though it is not the one that is active.
i.e. the current mode should be the selected one


https://user-images.githubusercontent.com/93143998/167392940-4b817358-7626-4f39-8665-225b287504e8.mp4

ichthus1604 commented:
@MaxHillebrand 

> Automatic coinjoin on/off should be on the first page, I think.

This was placed in the Customize dialog for consistency with the `Add new wallet` experience.

> Also the text in customize is still horrible, we gotta rewrite to make it more understandable.

Can you suggest the wording?

> Auto coinjoin threshold has the blue line on the right, not left, 

Fixed in latest commit.

> the number box looks different than the others

Fixed in latest commit.

ichthus1604 commented:
@MarnixCroes 

> When using Cancel it should not keep a previous selected item selected.

Fixed in latest commit.

soosr commented:
Sorry for the late answer but the `Cancel` button should have been disabled instead. As master suggest `NextCommand = CancelCommand;` so it was built to save immediately and to avoid future complexity for the new wallet settings (to store every value in a temporary, then save them) I would revert the last commit and just disable the cancel button.

ichthus1604 commented:
@soosr 

> so it was built to save immediately

Restored this behavior, now `NextCommand == CancelCommand`

soosr commented on WalletWasabi.Fluent/ViewModels/CoinJoinProfiles/CoinJoinProfilesViewModel.cs:
```diff
@@ -14,23 +15,44 @@ namespace WalletWasabi.Fluent.ViewModels.CoinJoinProfiles;
 		};
 
-		_selectedProfile = speedyProfile;
+		_selectedProfile =
+			Profiles.Where(x => x.MinAnonScoreTarget == keyManager.MinAnonScoreTarget)
+					.Where(x => x.MaxAnonScoreTarget == keyManager.MaxAnonScoreTarget)
+					.Where(x => x.FeeRateMedianTimeFrameHours == keyManager.FeeRateMedianTimeFrameHours)
+					.FirstOrDefault();
+
+		if (_selectedProfile == null && selectDefaultProfile)
+		{
+			_selectedProfile = Profiles[1];
+		}
+
+		if (_selectedProfile == null)
+		{
+			SelectedManualProfile = new ManualCoinJoinProfileViewModel(keyManager.AutoCoinJoin, keyManager.MinAnonScoreTarget, keyManager.MaxAnonScoreTarget, keyManager.FeeRateMedianTimeFrameHours);
+		}
+
+		this.WhenAnyValue(x => x.SelectedProfile, x => x.SelectedManualProfile)
+			.Subscribe(_ => ApplyChanges());
```
- It is unnecessary to set a manual coinjoin profile because if the user changes the selection it will be automatically created.
- Skip added to avoid unnecessary file save, because every time when the dialog was opened, it saved the wallet immediately, it shouldn't happen, mainly when we are in the wallet creation process. If the wallet got saved but the user presses the cancel, next when they open the software a nonwanted wallet will appear!

```suggestion
		var defaultProfile = Profiles[1];
		var currentProfile =
			Profiles.Where(x => x.MinAnonScoreTarget == keyManager.MinAnonScoreTarget)
				.Where(x => x.MaxAnonScoreTarget == keyManager.MaxAnonScoreTarget)
				.Where(x => x.FeeRateMedianTimeFrameHours == keyManager.FeeRateMedianTimeFrameHours)
				.FirstOrDefault()
			?? defaultProfile;

		_selectedProfile = selectDefaultProfile ? defaultProfile : currentProfile;

		this.WhenAnyValue(x => x.SelectedProfile, x => x.SelectedManualProfile)
			.Skip(1)
			.Subscribe(_ => ApplyChanges());
```

ichthus1604 replied:
@soosr 

> It is unnecessary to set a manual coinjoin profile because if the user changes the selection it will be automatically created.

This is done to show the current values in the dialog, otherwise you get bad UX because it seems to the user as if values were not really saved. For instance:

 - Set MinAnonScoreTarget to 20
 - click Done
 - close settings
 - re-open settings
 - you don't see the 20

soosr replied:
You are absolutely right, I forgot about when they use it from the wallet settings.

ichthus1604 commented on WalletWasabi.Fluent/Views/Dialogs/ManualCoinJoinProfileDialogView.axaml:
```diff
@@ -13,26 +13,51 @@
+                              IsApproximate="False"
+                              DockPanel.Dock="Bottom"
+                              Margin="10 -10 0 0" Classes="standalone" />
+          <Slider Minimum="0" Maximum="4" Value="{Binding PlebStopThresholdFactor}" Margin="10 0 0 0" VerticalAlignment="Center"
+                  TickFrequency="1" IsSnapToTickEnabled="True" />
```
@soosr 

> Slider is just nonsense here.

Might have been a misunderstanding on my part, but we discussed it with @danwalmsley and we agreed on this design.
Do you suggest to remove the slider and keep the currency input alone?

> Should the pleb stop really be under the customize?

Same as above. The idea is that only the simple one-click UI was visible first-hand, and that all other controls would be available under the "advanced" options UI. I can of course move it back to the first screen if needed, but I don't think it would really fit there with the current design

soosr replied:
When I tried out the slider it felt totally useless and my first impression was that it wouldn't be helping anyone.

> The idea is that only the simple one-click UI was visible first-hand, and that all other controls would be available under the "advanced" options UI.

ACK for the UX. It was just weird because it's under the `customize` section, but the profiles don't really change the pleb threshold.

soosr commented:
0 pleb threshold has fullstop in it for no reason.
![image](https://user-images.githubusercontent.com/16364053/168242506-4c200d76-b736-4c3a-bcdf-b4d5f86428ed.png)

Bug:
1. Type 0 to the pleb threshold.
2. Press ENTER.
3. Customize dialog cannot be open anymore.

ichthus1604 commented:
@soosr 

> 1. Type 0 to the pleb threshold.

Should we prevent that? Is zero a valid value here?

soosr commented:
> @soosr
> 
> > 1. Type 0 to the pleb threshold.
> 
> Should we prevent that? Is zero a valid value here?

Nope, that is a valid value. The issue is that the dialog cannot open anymore.

ichthus1604 commented:
@soosr Fixed zero value issues

ichthus1604 commented on WalletWasabi.Fluent/ViewModels/CoinJoinProfiles/CoinJoinProfilesViewModel.cs:
```diff
@@ -44,26 +66,39 @@ public CoinJoinProfilesViewModel(KeyManager keyManager)
+		}
 
-		Navigate().To(new AddedWalletPageViewModel(keyManager));
+		_keyManager.ToFile();
```
@soosr fixed in lastest commit

ichthus1604 commented:
@danwalmsley @soosr @MaxHillebrand 

Latest commit: 

 - Merged master, which had a change that removed one of the sliders
 - Moved Auto coinjoin switch to main settings screen, while keeping the one inside the inner dialog for new wallets only.

MaxHillebrand reviewed (dismissed):
ACK, well done.

ichthus1604 commented:
Latest commit contains 2 cosmetic fixes:

 - fixed dropshadow clipping on Coinjoin Profile Selector
 - Made Coinjoin Profile Selector height smaller in the settings UI to prevent vertical scrolling:

![image](https://user-images.githubusercontent.com/98904713/169832553-3d26cea3-bae3-4798-aa10-38399cab1298.png)

MarnixCroes commented:
I´m thinking, it might be an improvement if there is a dropdown for the explainers of the profiles?
The explainers are not really needed to be always displayed imo, `Minimize Costs`, `Maximize Speed` & `Maximize Privacy` speak for themselves most of the time/for most users.

ichthus1604 commented:
Latest commit improves layout:

![image](https://user-images.githubusercontent.com/98904713/170321071-f8a85fca-0eb3-4855-a846-34b1a86831de.png)

danwalmsley commented:
IMO this PR got too complex for that its doing.

Id like to close it and we do another PR with a simple solution. Simply make the exisiting Profiles screen a DialogViewModel that returns a result.

That way there will be little to no ui changes and the dialog can be shared.

Wallet settings screen would simply say something like:

Selected Profile: Economical 
[ Select Profile ]

Clicking that button would simply bring up the dialog like it normally would.

Everything else then remains the same as it was.
